pkg/networkservice/ipam/singlepointipam/range.go

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) 2022 Nordix and its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package singlepointipam
+
+import (
+	"net"
+
+	"github.com/pkg/errors"
+)
+
+// IpamRange - range of IP addresses
+type IpamRange struct {
+	RangeStart net.IP
+	RangeEnd   net.IP
+}
+
+// IpamNet - IP network with optional IP ranges to limit address scope
+type IpamNet struct {
+	Network *net.IPNet
+	Ranges  []*IpamRange // optional: limit IpamNet by specifying valid range(s) within Network
+}
+
+// NewIpamNet - creates a new IpamNet
+func NewIpamNet() *IpamNet {
```

Comment:
@zolug 

Seems like it can be replaced with https://github.com/networkservicemesh/sdk/blob/main/pkg/tools/ippool/ippool.go#L37

IPPool already supports ranges/contains operation. The data structure is based on a red-black tree and should be extremely fast.

Am I missing something?

Reply:
Hi!

For context here's the Remote VLAN NSE PR with the configuration changes that we would like to achieve:
https://github.com/networkservicemesh/cmd-nse-remote-vlan/pull/111

IpamNet is mostly aimed to be used verifying input in the Remote Vlan NSE (already relies on net pkg).

Correct me, if i'm wrong but IMHO ippool could be only used partially (without maybe major changes). For example to replace `IpamRange` slice in `IpamNet`.
That's because in our case the IPAM must be able to handle IPs of multiple ranges that all belong to the same pre-configured subnet, and not just a "random" set of IP addresses.
Meaning that for example the pre-configured network prefix had to be still stored some place, in order to execute admittance checks of ranges. (But the returned IPs are also assigned using the pre-configured network prefix.)

Thus for me it seems that e.g. an additional ippool could be needed for admittance control, that would contain the whole pre-configured subnet, but it wouldn't be used to allocate IPs from. (And as it seems the ippool package would also require a `ContainsRange` function but that's a minor thing.) And I still might struggle with some config checks.

On the other hand, I guess the allowed IP ranges could be mapped to a net + exclude ranges but IMHO that would add unnecessary complexity.

So my point is that IMHO a new struct still would be required even if built on top of ippool.

Maybe sg like this:
```
type IpamNet struct {
	Network *net.IPNet
	Ranges *ippool.IPPool
}
```

Or did I misundertand sg?

Reply:
Hello @zolug !

I started looking deeply into the patch and found that the new functionality is not covered by the unit test for the chain element.

Can we add a test? I think it could explain the necessity of the new `IpamNet`